docs/design-v1.md

```diff
@@ -0,0 +1,106 @@
+# Caching Service Design Doc
+
+## Purpose
+
+Provide a generalized cache store that can be used from any other service, allowing users to avoid re-running expensive computation that has already been completed for the same dataset.
+
+## MVP Specifications
+
+**_Set a cache value_**
+
+To set a cache value, send a POST request with content-type "multipart/form-data" (to facilitate the file upload).
+
+The fields are:
+
+* `service` - required - name of the service
+* `token` - required - authentication token for the service
+* `params` - required - an arbitrary set of data used in the cache's namespace
+* `file` - required - multipart form-data file to cache
+* `expiration_days` - optional - total time in days until the cache's file gets deleted
+
+```
+curl -i -X POST -H "Content-Type: multipart/form-data"
+  -F file=@/home/u/bio.zip
+  -F service=MyService
+  -F token=xyz
+  -F params="{..arbitrary_json_data..}"
+  -F expiration_days=30
+  https://<caching_service_url>/v1/set
+```
+
+If the cache succeeds, you will get back an empty 200 response.
+
+If your auth token is invalid, you will receive a 403. Otherwise, there should be no errors.
+
+**_Request a cache value_**
+
+To fetch a cache value, make a POST request to `/v1/get` with content-type
+"application/x-www-form-urlencoded" and form fields for "service", "token", and "params".
+
+```
+curl -i -X POST -H "Content-Type: application/x-www-form-urlencoded"
+  -F service=MyService
+  -F token=xyz
+  -F params="{..arbitrary_json_data..}"
+  https://<caching_service_url>/v1/get
+```
+
+The service token and params JSON must exactly match those that you used when setting the cache in order to get the same file back.
+
+If the cache file is found, then the server will response with a 200 and the URL of the cached file. Otherwise, the response will have status 404 with an empty body. A 403 response indicates that the service token is invalid. 
```

Comment:
I'm missing something here - it returns the URL of the cached file? Why wouldn't it just return the file?

Reply:
Just so they can manage when/how they want to download the file themselves, especially if it's very large. Might not be useful though, I can just put the file in the response -- whatever is simplest.

Reply:
> Just so they can manage when/how they want to download the file themselves

I assumed they'd do that by deciding when to call the service. I'm not grokking the benefit of making a call to url1 which then just tells you to call url2.